File: backend/app/models/receivable.py
from datetime import datetime
from decimal import Decimal
from typing import Optional

from sqlalchemy import DateTime, ForeignKey, Numeric, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base


class Receivable(Base):
    __tablename__ = "receivables"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    number: Mapped[Optional[str]] = mapped_column(String(40), unique=True, index=True)
    sale_id: Mapped[Optional[int]] = mapped_column(ForeignKey("sales.id", ondelete="SET NULL"), index=True)
    client_id: Mapped[Optional[int]] = mapped_column(ForeignKey("clients.id", ondelete="SET NULL"), index=True)
    description: Mapped[str] = mapped_column(String(220), nullable=False)
    original_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    paid_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    balance_amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False, default=0)
    status: Mapped[str] = mapped_column(String(30), nullable=False, default="open", index=True)
    due_date: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
    settled_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True))
    notes: Mapped[Optional[str]] = mapped_column(Text)

    payments = relationship(
        "ReceivablePayment",
        back_populates="receivable",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
    client = relationship("Client", lazy="selectin")
    sale = relationship("Sale", lazy="selectin")

    @property
    def client_name(self) -> str | None:
        return self.client.name if self.client is not None else None

    @property
    def sale_number(self) -> str | None:
        return self.sale.number if self.sale is not None else None

    @property
    def sale_sold_at(self) -> datetime | None:
        return self.sale.sold_at if self.sale is not None else None

    @property
    def sale_items(self) -> list[dict[str, object]]:
        if self.sale is None:
            return []
        return [
            {
                "id": item.id,
                "description": item.description,
                "quantity": item.quantity,
                "unit": item.unit,
                "unit_price": item.unit_price,
                "total_price": item.total_price,
            }
            for item in self.sale.items
        ]


class ReceivablePayment(Base):
    __tablename__ = "receivable_payments"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    receivable_id: Mapped[int] = mapped_column(ForeignKey("receivables.id", ondelete="CASCADE"), index=True)
    user_id: Mapped[Optional[int]] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    amount: Mapped[Decimal] = mapped_column(Numeric(12, 2), nullable=False)
    method: Mapped[str] = mapped_column(String(30), nullable=False, default="dinheiro")
    notes: Mapped[Optional[str]] = mapped_column(Text)
    paid_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )

    receivable = relationship("Receivable", back_populates="payments")

File: admin_app/admin_flutter/lib/models/receivable.dart
class Receivable {
  const Receivable({
    required this.id,
    required this.description,
    required this.originalAmount,
    required this.paidAmount,
    required this.balanceAmount,
    required this.status,
    required this.createdAt,
    required this.saleItems,
    required this.payments,
    this.number,
    this.saleId,
    this.clientId,
    this.clientName,
    this.saleNumber,
    this.saleSoldAt,
    this.dueDate,
    this.settledAt,
    this.notes,
    this.entryType = 'legacy',
    this.fiscalDocumentId,
    this.fiscalDocumentStatus,
    this.fiscalDocumentType,
    this.fiscalDocumentSeries,
    this.fiscalDocumentNumber,
  });

  final int id;
  final String? number;
  final int? saleId;
  final int? clientId;
  final String? clientName;
  final String? saleNumber;
  final DateTime? saleSoldAt;
  final String description;
  final double originalAmount;
  final double paidAmount;
  final double balanceAmount;
  final String status;
  final DateTime? dueDate;
  final DateTime createdAt;
  final DateTime? settledAt;
  final String? notes;
  final String entryType;
  final int? fiscalDocumentId;
  final String? fiscalDocumentStatus;
  final String? fiscalDocumentType;
  final int? fiscalDocumentSeries;
  final int? fiscalDocumentNumber;
  final List<ReceivableSaleItem> saleItems;
  final List<ReceivablePayment> payments;

  factory Receivable.fromJson(Map<String, dynamic> json) {
    return Receivable(
      id: json['id'] as int,
      number: json['number'] as String?,
      saleId: json['sale_id'] as int?,
      clientId: json['client_id'] as int?,
      clientName: json['client_name'] as String?,
      saleNumber: json['sale_number'] as String?,
      saleSoldAt: json['sale_sold_at'] == null
          ? null
          : DateTime.parse(json['sale_sold_at'] as String).toLocal(),
      description: json['description'] as String,
      originalAmount: _toDouble(json['original_amount']),
      paidAmount: _toDouble(json['paid_amount']),
      balanceAmount: _toDouble(json['balance_amount']),
      status: json['status'] as String? ?? 'open',
      dueDate: json['due_date'] == null
          ? null
          : DateTime.parse(json['due_date'] as String).toLocal(),
      createdAt: DateTime.parse(json['created_at'] as String).toLocal(),
      settledAt: json['settled_at'] == null
          ? null
          : DateTime.parse(json['settled_at'] as String).toLocal(),
      notes: json['notes'] as String?,
      entryType: json['entry_type'] as String? ?? 'legacy',
      fiscalDocumentId: json['fiscal_document_id'] as int?,
      fiscalDocumentStatus: json['fiscal_document_status'] as String?,
      fiscalDocumentType: json['fiscal_document_type'] as String?,
      fiscalDocumentSeries: json['fiscal_document_series'] as int?,
      fiscalDocumentNumber: json['fiscal_document_number'] as int?,
      saleItems: ((json['sale_items'] as List<dynamic>?) ?? [])
          .map(
            (item) => ReceivableSaleItem.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
      payments: ((json['payments'] as List<dynamic>?) ?? [])
          .map(
            (item) => ReceivablePayment.fromJson(item as Map<String, dynamic>),
          )
          .toList(),
    );
  }
}

class ReceivableSaleItem {
  const ReceivableSaleItem({
    required this.id,
    required this.description,
    required this.quantity,
    required this.unit,
    required this.unitPrice,
    required this.totalPrice,
  });

  final int id;
  final String description;
  final double quantity;
  final String unit;
  final double unitPrice;
  final double totalPrice;

  factory ReceivableSaleItem.fromJson(Map<String, dynamic> json) {
    return ReceivableSaleItem(
      id: json['id'] as int,
      description: json['description'] as String? ?? '-',
      quantity: _toDouble(json['quantity']),
      unit: json['unit'] as String? ?? 'un',
      unitPrice: _toDouble(json['unit_price']),
      totalPrice: _toDouble(json['total_price']),
    );
  }
}

class ReceivablePayment {
  const ReceivablePayment({
    required this.id,
    required this.receivableId,
    required this.amount,
    required this.method,
    required this.paidAt,
    this.userId,
    this.userName,
    this.notes,
    this.reversedAt,
    this.reversedByUserId,
    this.reversedByUserName,
    this.reversalReason,
  });

  final int id;
  final int receivableId;
  final int? userId;
  final String? userName;
  final double amount;
  final String method;
  final String? notes;
  final DateTime? reversedAt;
  final int? reversedByUserId;
  final String? reversedByUserName;
  final String? reversalReason;
  bool get isReversed => reversedAt != null;
  final DateTime paidAt;

  factory ReceivablePayment.fromJson(Map<String, dynamic> json) {
    return ReceivablePayment(
      id: json['id'] as int,
      receivableId: json['receivable_id'] as int,
      userId: json['user_id'] as int?,
      userName: json['user_name'] as String?,
      amount: _toDouble(json['amount']),
      method: json['method'] as String? ?? 'dinheiro',
      notes: json['notes'] as String?,
      reversedAt: json['reversed_at'] == null
          ? null
          : DateTime.parse(json['reversed_at'] as String).toLocal(),
      reversedByUserId: json['reversed_by_user_id'] as int?,
      reversedByUserName: json['reversed_by_user_name'] as String?,
      reversalReason: json['reversal_reason'] as String?,
      paidAt: DateTime.parse(json['paid_at'] as String).toLocal(),
    );
  }
}

class ReceivablePaymentPayload {
  const ReceivablePaymentPayload({
    required this.amount,
    required this.method,
    this.notes,
  });

  final double amount;
  final String method;
  final String? notes;

  Map<String, dynamic> toJson() => {
    'amount': amount,
    'method': method,
    'notes': notes?.trim().isEmpty == true ? null : notes?.trim(),
  };
}

class ReceivableManualPayload {
  const ReceivableManualPayload({
    required this.clientId,
    required this.amount,
    required this.description,
    this.dueDate,
    this.notes,
    this.entryType = 'legacy',
  });

  final int clientId;
  final double amount;
  final String description;
  final DateTime? dueDate;
  final String? notes;
  final String entryType;

  Map<String, dynamic> toJson() => {
    'client_id': clientId,
    'amount': amount,
    'description': description.trim().isEmpty
        ? 'Lançamento manual de crediário'
        : description.trim(),
    'due_date': dueDate?.toIso8601String(),
    'notes': notes?.trim().isEmpty == true ? null : notes?.trim(),
    'entry_type': entryType,
  };
}

double _toDouble(Object? value) {
  if (value == null) return 0;
  if (value is num) return value.toDouble();
  return double.tryParse(value.toString()) ?? 0;
}
